pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -391,43 +398,60 @@ type podVolume struct {
 // the volume spec and put it in both the actual and desired state of worlds. If no running
 // container is mounting the volume, the volume will be removed by desired state of world's populator and
 // cleaned up by the reconciler.
-func (rc *reconciler) reconstructStates(podsDir string) {
+func (rc *reconciler) syncStates(podsDir string) {
 	// Get volumes information by reading the pod's directory
 	podVolumes, err := getVolumesFromPodDir(podsDir)
 	if err != nil {
 		glog.Errorf("Cannot get volumes from disk %v", err)
 		return
 	}
+
+	volumesNeedUpdate := make(map[api.UniqueVolumeName]*operationexecutor.VolumeToMount)
 	for _, volume := range podVolumes {
 		volumeToMount, err := rc.reconstructVolume(volume)
 		if err != nil {
 			glog.Errorf("Could not construct volume information: %v", err)
 			continue
 		}
-
 		// Check if there is an pending operation for the given pod and volume.
 		// Need to check pending operation before checking the actual and desired
 		// states to avoid race condition during checking. For example, the following
 		// might happen if pending operation is checked after checking actual and desired states.
 		// 1. Checking the pod and it does not exist in either actual or desired state.
 		// 2. An operation for the given pod finishes and the actual state is updated.
 		// 3. Checking and there is no pending operation for the given pod.
-		if rc.operationExecutor.IsOperationPending(volumeToMount.VolumeName, volumeToMount.PodName) {
-			continue
-		}
-		desiredPods := rc.desiredStateOfWorld.GetPods()
-		actualPods := rc.actualStateOfWorld.GetPods()
-		if desiredPods[volume.podName] || actualPods[volume.podName] {
+		// During state reconstruction period, no new volume operations could be issued. If the
+		// mounted path is not in either pending operation, or actual or desired states, this
+		// volume needs to be reconstructed back to the states.
+		pending := rc.operationExecutor.IsOperationPending(volumeToMount.VolumeName, volumeToMount.PodName)
+		dswExist := rc.desiredStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
+		aswExist, _, _ := rc.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
+
+		// In case this is the first time to reconstruct state since kubelet starts, if the
+		// mounted path is for a persistant volume, it must have been mounted before kubelet
+		// restarts because no mount operations for persistent volumes could be started at this time (node
+		// status has not yet been updated before first syncStates finishes), In this case, the volume state
+		// should be put back to actual state no matter desired state has it or not.
+		if rc.SyncStatesStarted() {
```

Comment:
It is not always return true. Only after the first SyncStates finishes, it will update the time. During first time SyncStates is executed, it returns false.

But somehow I think there is typo here, it should be !SyncStatesStarted(). The reason I put this check is because, without this check when kubelet restarts, the following events might happen
1. pod manager gets existing pods and add volume info into desired state
2. syncStates process scans existing directory, it checks the volume is already in desired state, skip it 
3. After the first SyncState finishes, kubelet node status updater will update the list of mounted volumes and the skipped volume from step 2 will not in the list
3. pod is deleted during kubelet restarts, master gets the list of mounted volume which does not include the volume so check for detach will pass for this volume which causes detaching while mounting problem